### PR TITLE
Improve thread safety of LeaderElection algorithm

### LINKED ISSUE #59: LeaderElectionBase is not thread safe
**Describe the bug**

`LeaderElectionBase` is not thread safe.

- If two threads call `StartAsync()` at the same time on a stopped LeaderElection instance, then it is possible for both threads to create a new `_leadershipLoopCancellationTokenSource` and `_leadershipLoopTask`. The first assigned task will be orphaned by the second assignment, and that orphaned task will execute in the background for the lifetime of the application. 

- `StopAsync()` suffers from a **Check-then-Act** vulnerability, specifically inside the "if field not null" expressions. Two threads/tasks could enter these blocks at the same time, then one might pause while the other completes the function and set the member to null, resulting in a null reference exception when the other thread/task resumes.

- The public `TryAcquireLeadershipAsync()` method contains a potential deadlock. This method invokes the `LeadershipChanged` event while the `_leadershipSemaphore` is acquired. If the user's callback was to somehow call `TryAcquireLeadershipAsync()` again, then a deadlock will occur.

**Expected behavior**

- [ ] The state of a LeaderElection instance should be thread safe.
  - In the example given above, the first thread should block the second thread until it completes all work on the internal state.
  - Read-only properties like `IsLeader` and `LastLeadershipRenewal` do not require the critical section as long as the code is written such that these fields can be safely accessed outside of the critical section, by strategic use of Volatile and Interlocked operations.
- [ ] Callback events should be invoked **outside** of the critical section to prevent deadlocks and other potential reentrancy issues. They should be executed **after** all internal state has been updated.
- [ ] Derived implementations should not _need_ to worry about thread-safety. In other words, the base implementation should guarantee that the protected virtual methods are called within a thread-safe context. This implies that derived implementations **must not** fire Events.

### LINKED ISSUE #60: LeaderElectionBase.StartAsync() links the leader-loop task to the supplied cancellation token
**Describe the bug**

The `LeaderElectionBase.StartAsync(CancellationToken)` method links the `_leadershipLoopCancellationTokenSource` to the provided cancellation token. This allows the caller to cancel the leader-loop task without the LeaderElection instances' knowledge. This is wrong. Unless explicitly documented otherwise, the provided cancellation token should only be used to cancel the `StartAsync()` method itself. 

The user must call `StopAsync()` or `DisposeAsync()` to cancel the leader-loop task.

There is also no reason to expect that the provided cancellation token is long-lived (it isn't documented anywhere). In many cases tokens are short-lived, such as those created when processing an ASP.NET API request, which is disposed as soon as the request completes.

**Expected behavior**

- [ ] The provided cancellation token is only used to cancel the `StartAsync()` method itself.
- [ ] The provided cancellation token has no effect on the lifetime of the leader-loop task.

### PR DESCRIPTION
**Pull request by mrfootoyou**
## Description

> [!IMPORTANT]
> This PR contains functional changes. See the review comments for details. 

This pull request refactors `BaseLeaderElection` to ensure robust concurrent access to leadership status and the background loop by replacing volatile fields with explicit Volatile and Interlocked operations. This change prevents potential race conditions and torn reads, particularly for leadership renewal timestamps, while centralizing state transitions to maintain consistency across the lifecycle of the leader election instance. This resolves #59

It resolves #60 by **not** tying the `_leaderLoopTaskCTS` to the `StartAsync()` cancellation token.

It also introduces several improvements and clarifications to the leader election interface, its documentation, and improves test coverage. It includes enhanced XML documentation for the `ILeaderElection` interface, the addition of new properties and exception details, and a significant expansion of unit tests to cover edge cases and error handling. Additionally, some package dependencies have been updated.

---

## Checklist

- [x] My code follows the project's coding style
- [x] I have self-reviewed my changes
- [x] I have added tests (or explain why not needed)
- [ ] I have updated the documentation
- [x] I have linked relevant issue(s)
- [x] I have included meaningful commit messages

---

## Related Issues

Closes #59 and #60

---

## Additional Notes

Note that the target branch is `refactor/di-extensions` and should not be merged until after #64 is merged.

### CODE REVIEW

File: src/LeaderElection/LeaderElectionBase.cs [hunk 1]
```diff
@@ -28,238 +51,449 @@ protected LeaderElectionBase(
...
+        ThrowIfDisposed();
+
+        // Fast checks before acquiring the semaphore to avoid unnecessary waits
+        if (!LeaderLoopRunning)
         {
-            LogStoppingLeaderElectionForInstanceInstanceid(_settings.InstanceId);
+            throw new InvalidOperationException(
+                "Leader loop must be running to acquire leadership"
+            );
+        }
```
Comment (mrfootoyou) [inline src/LeaderElection/LeaderElectionBase.cs]: ❗This is a functional change: `TryAcquireLeadershipAsync()` will not acquire leadership unless the leader-loop is running. This makes sense to me because without a running leader-loop the instance will not know if it loses leadership, resulting in multiple leaders.

File: src/LeaderElection/ILeaderElection.cs [hunk 2]
```diff
@@ -23,39 +35,56 @@ public interface ILeaderElection : IAsyncDisposable
...
     /// <summary>
-    /// Attempts to acquire leadership manually
+    /// Attempts to acquire leadership manually.
     /// </summary>
+    /// <remarks>
+    /// Must be called while the leader loop is running.
+    /// </remarks>
     /// <param name="cancellationToken">Cancellation token</param>
     /// <returns>True if leadership was acquired, false otherwise</returns>
+    /// <exception cref="InvalidOperationException">Thrown if the leader loop is not running.</exception>
```
Comment (mrfootoyou) [inline src/LeaderElection/ILeaderElection.cs:67]: ❗This is a functional change: `TryAcquireLeadershipAsync()` will not acquire leadership unless the leader-loop is running. This makes sense to me because without a running leader-loop the instance will not know if it loses leadership, resulting in multiple leaders.

File: src/LeaderElection/ILeaderElection.cs [hunk 3]
```diff
@@ -23,39 +35,56 @@ public interface ILeaderElection : IAsyncDisposable
...
     Task<bool> TryAcquireLeadershipAsync(CancellationToken cancellationToken = default);
 
     /// <summary>
-    /// Executes a task only if this instance is the leader
+    /// Executes a task only if this instance is the leader.
     /// </summary>
     /// <param name="task">The task to execute</param>
     /// <param name="cancellationToken">Cancellation token</param>
     /// <returns>Task representing the async operation</returns>
+    /// <exception cref="ArgumentNullException">Thrown if the task is null.</exception>
```
Comment (mrfootoyou) [inline src/LeaderElection/ILeaderElection.cs:77]: ❗This is a functional change: `RunTaskIfLeaderAsync()` will throw if the supplied task is null. The parameter isn't nullable so it is unlikely anyone will even notice this change.

File: src/LeaderElection/ILeaderElection.cs [hunk 4]
```diff
@@ -23,39 +35,56 @@ public interface ILeaderElection : IAsyncDisposable
...
+    /// <exception cref="ObjectDisposedException">Thrown if the leader election instance has been disposed.</exception>
     Task RunTaskIfLeaderAsync(Func<Task> task, CancellationToken cancellationToken = default);
```
Comment (mrfootoyou) [inline src/LeaderElection/ILeaderElection.cs:79]: This should probably return a value indicating if the task was executed or not.

It would be nice if this method passed a cancellation token to the Func that cancels if leadership is lost while it is running. This would enable the Func to be a long-running task. Q: Should this "leadership token" be tied to the passed in token? I think so since there is no other reason to pass in a cancellation token other than to cancel the task.

File: src/LeaderElection/LeaderElectionSettingsBase.cs [hunk 5]
```diff
@@ -15,17 +15,57 @@ public class LeaderElectionSettingsBase
...
     /// Must be greater than or equal to zero.
     /// Default is 5 seconds.
     /// </summary>
+    /// <remarks>
+    /// Unexpected errors during leader acquisition will exponentially increase this interval by
+    /// the retry backoff factor. The computed retry interval will be randomly adjusted by ± the
+    /// jitter percentage, up to the maximum retry interval. The exact formula is:
+    /// <code>
+    /// retry := min(RetryInterval * (RetryBackoffFactor ^ errorCount) * (1 ± rand(RetryJitter)), MaxRetryInterval)
+    /// </code>
```
Comment (mrfootoyou) [inline src/LeaderElection/LeaderElectionSettingsBase.cs:29]: ❗This is functional change: Deprecated `MayRetryAttempts` in favor of more flexible retry policy with jitter, which is important when many instance are contending for the same resource, exponential backoff when errors occur, and a maximum value.

File: src/LeaderElection/LeaderElectionSettingsBase.cs [hunk 6]
```diff
@@ -15,17 +15,57 @@ public class LeaderElectionSettingsBase
...
+    /// Default is 0.10 (±10%).
+    /// </summary>
+    [Range(0.0, 1.0)]
+    public double RetryJitter { get; set; } = 0.10;
+
+    /// <summary>
+    /// The maximum retry interval (after applying the backoff factor and jitter) when acquiring leadership.
+    /// <para/>
+    /// Must be greater than or equal to <see cref="RetryInterval"/>.
+    /// Default is 40 seconds.
```
Comment (mrfootoyou) [inline src/LeaderElection/LeaderElectionSettingsBase.cs:61]: 40 is a weird number but it is based on the existing implementation: (RetryInterval * 2^MayRetryAttempts) => (5 * 2^3) = 40

File: src/LeaderElection/LeaderElectionSettingsBase.cs [hunk 7]
```diff
@@ -41,14 +81,16 @@ public class LeaderElectionSettingsBase
...
     /// The maximum number of retry attempts before giving up on acquiring leadership.
     /// <para/>
-    /// Must be greater than or equal to zero.
     /// Default is 3.
     /// </summary>
-    [Range(0, int.MaxValue)]
+    [Obsolete(
+        "This setting is no longer used and will be removed in a future version. Use MaxRetryInterval instead."
+    )]
     public int MaxRetryAttempts { get; set; } = 3;
```
Comment (mrfootoyou) [inline src/LeaderElection/LeaderElectionSettingsBase.cs:89]: Deprecate `MayRetryAttempts`. This is a misleading property since the app does not stop trying to acquire leadership.

File: src/LeaderElection/LeaderElectionBase.cs [hunk 8]
```diff
@@ -19,247 +46,498 @@ public abstract partial class LeaderElectionBase<TSettings> : ILeaderElection
...
+            cts?.Dispose();
+        }
+    }
 
-        await Task.CompletedTask.ConfigureAwait(false);
+    /// <inheritdoc />
+    public virtual Task StopAsync(CancellationToken cancellationToken = default)
+    {
+        ThrowIfDisposed();
+        return InternalStopAsync(true, cancellationToken);
```
Comment (mrfootoyou) [inline src/LeaderElection/LeaderElectionBase.cs:144]: ❗This is a functional change: `InternalStopAsync(true)` means "release leadership". Previously, leadership was only released when `_settings.EnableGracefulShutdown` is true. Now, this setting is only used when disposing, i.e., during a non-explicit "shutdown".

Fortunately, the default value of `EnableGracefulShutdown` is true so people using the defaults will not notice this change. (I can't imagine any good reason to set it false.)

We should consider adding an optional "releaseLeadership" parameter to `StopAsync()`.

File: src/LeaderElection/LeaderElectionBase.cs [hunk 9]
```diff
@@ -19,247 +46,498 @@ public abstract partial class LeaderElectionBase<TSettings> : ILeaderElection
...
         ThrowIfDisposed();
 
-        var retryCount = 0;
+        // Fast checks before acquiring the semaphore to avoid unnecessary waits
+        if (!LeaderLoopRunning)
+        {
+            throw new InvalidOperationException(
+                "Leader loop must be running to acquire leadership"
+            );
+        }
```
Comment (mrfootoyou) [inline src/LeaderElection/LeaderElectionBase.cs:158]: ❗This is a functional change: `TryAcquireLeadershipAsync()` will not acquire leadership unless the leader-loop is running. This makes sense to me because without a running leader-loop the instance will not know if it loses leadership, resulting in multiple leaders.

File: src/LeaderElection/LeaderElectionBase.cs [hunk 10]
```diff
@@ -19,247 +46,498 @@ public abstract partial class LeaderElectionBase<TSettings> : ILeaderElection
...
+
+    /// <summary>
+    /// Tries to acquire leadership. Implementations should return true if leadership was
+    /// successfully acquired, or false if leadership could not be acquired (e.g. another
+    /// instance holds the leadership).
+    /// <para/>
+    /// If an unexpected error occurs, implementations should abandon leadership in their
+    /// internal state and throw an exception. The leader loop will catch the exception,
+    /// log it, and attempt to reacquire leadership on the next iteration. This is a safety
+    /// measure to avoid a situation where an instance thinks it is the leader when it is not.
```
Comment (mrfootoyou) [inline src/LeaderElection/LeaderElectionBase.cs:249]: This is an important implementation note: We are prioritizing correctness (at most one leader) over leader continuity. In other words, we prefer to have no leader (hopefully for a short period of time) rather than the possibility of multiple leaders due to some unforeseen error.

File: src/LeaderElection/LeaderElectionBase.cs [hunk 11]
```diff
@@ -19,247 +46,498 @@ public abstract partial class LeaderElectionBase<TSettings> : ILeaderElection
...
-                        LogLeadershipAcquiredForInstanceInstanceid(_settings.InstanceId);
-                        retryCount = 0;
+                        acquiredLeadership = true;
+                        LogLeadershipAcquired(_settings.InstanceId);
                     }
                     else
                     {
-                        // Very common that leadership acquisition fails because another instance holds the leadership - log at debug level in that case and avoid log noise
-                        LogFailedToAcquireLeadershipWillRetry();
-                        retryCount++;
```
Comment (mrfootoyou) [inline src/LeaderElection/LeaderElectionBase.cs:88]: ❗This is a functional change: Failure to acquire/renew leadership does not result in exponential backoff. These are not exceptional conditions.

File: src/LeaderElection/LeaderElectionBase.cs [hunk 12]
```diff
@@ -19,247 +46,498 @@ public abstract partial class LeaderElectionBase<TSettings> : ILeaderElection
...
+        else
+        {
+            var backoffFactor = Math.Pow(_settings.RetryBackoffFactor, Math.Min(errorCount, 30));
+            var jitter = _settings.RetryJitter - (_getRandomDouble() * _settings.RetryJitter * 2);
+            return TimeSpan.FromTicks(
+                Math.Min(
+                    (long)(_settings.RetryInterval.Ticks * backoffFactor * (1.0 - jitter)),
+                    _settings.MaxRetryInterval.Ticks
+                )
+            );
```
Comment (mrfootoyou) [inline src/LeaderElection/LeaderElectionBase.cs]: ❗This is a functional change: Retry with configurable jitter, exponential backoff, and maximum.

File: src/LeaderElection/LeaderElectionSettingsBase.cs (hunk 5 above)
Comment (gregyjames) [reply src/LeaderElection/LeaderElectionSettingsBase.cs:29]: I do like this solution better, one thing I have been thinking about is do you think it makes sense to manually define the logic for this in our library? Or to offload the resilience logic by adding support ResiliencePipelines and Microsoft.Extensions.Resilience? Then users could define their own pipelines and we can provide a default implementation when one is not defined from the settings? Would definitely reduce some of the complexity we add from doing this manually and be more flexible for end users.

File: src/LeaderElection/LeaderElectionBase.cs [hunk 13]
```diff
@@ -19,252 +46,518 @@ public abstract partial class LeaderElectionBase<TSettings> : ILeaderElection
...
 
-    private bool IsDisposed => Volatile.Read(ref _disposedValue) == 1;
-    public bool IsLeader => _isLeader && !IsDisposed;
-    public DateTime LastLeadershipRenewal => _lastLeadershipRenewalTime.UtcDateTime;
+    /// <inheritdoc />
+    public bool IsLeader => Volatile.Read(ref _isLeader);
 
+    /// <inheritdoc />
+    public DateTime LastLeadershipRenewal =>
+        IsLeader
```
Comment (gregyjames) [inline src/LeaderElection/LeaderElectionBase.cs:92]: Wouldn't this open up the potential for a race condition between two separate volatile/interlocked operations? Do you think it would be better check ticks first or create a state object for leadership so both are accessed/saved together?

File: src/LeaderElection/LeaderElectionBase.cs [hunk 14]
```diff
@@ -19,252 +46,518 @@ public abstract partial class LeaderElectionBase<TSettings> : ILeaderElection
...
-        );
-    }
-
-    public virtual Task StopAsync(CancellationToken cancellationToken = default)
-    {
-        ThrowIfDisposed();
-        return InternalStopAsync(cancellationToken);
+        else
+        {
+            var backoffFactor = Math.Pow(_settings.RetryBackoffFactor, Math.Min(errorCount, 30));
```
Comment (gregyjames) [inline src/LeaderElection/LeaderElectionBase.cs]: nit: Can we move the 30 here to a named constant?

**Review by gregyjames — COMMENTED** [review]
Quality work as always, agreed on the presented tradeoffs and functionality changes and prioritizing correctness. A few comments and general thoughts.

File: src/LeaderElection/ILeaderElection.cs (hunk 4 above)
Comment (gregyjames) [reply src/LeaderElection/ILeaderElection.cs:79]: Agreed, passing in a cancellation token would be a nice touch and agreed it should be linked to the leadership token since it should be cancelled if the instance is no longer the leader.

File: src/LeaderElection/LeaderElectionSettingsBase.cs (hunk 5 above)
Comment (mrfootoyou) [reply src/LeaderElection/LeaderElectionSettingsBase.cs:29]: I thought about that too but was hesitant to add another dependency. I created issue #69 so we can think about it later.

File: src/LeaderElection/LeaderElectionBase.cs (hunk 13 above)
Comment (mrfootoyou) [reply src/LeaderElection/LeaderElectionBase.cs:92]: It does, but I'm pretty sure it doesn't matter because:
- Both of these values are written at the same instance.
- Both are written at the very end of the mutex.
- `_isLeader` is written **last**.

So, if this property executes after ticks is updated but before `_isLeader`, then these are the possible outcomes:

| `_isLeader` change | outcome 
| ---------      | -------
| false -> false (no change) | min time (correct)
| false -> true  | min time (technically correct)
| true -> false  | old time (correct - ticks not updated)
| true -> true (no change)   | new time (correct)

The only reason we even need the `Interlocked.Read` is to avoid tearing on 32-bit processors (same reason for "ticks" instead of DateTime). If it wasn't for that, the Volatile ops on _isLeader would be sufficient.

File: src/LeaderElection/LeaderElectionBase.cs [hunk 15]
```diff
@@ -19,252 +46,584 @@ public abstract partial class LeaderElectionBase<TSettings> : ILeaderElection
...
-        return InternalStopAsync(cancellationToken);
+        else
+        {
+            const int maxExponent = 30; // to prevent overflow
+            var backoffFactor = Math.Pow(
+                _settings.RetryBackoffFactor,
+                Math.Min(errorCount, maxExponent)
+            );
+            var jitter = _settings.RetryJitter - (_getRandomDouble() * _settings.RetryJitter * 2);
+            return TimeSpan.FromTicks(
```
Comment (gregyjames) [inline src/LeaderElection/LeaderElectionBase.cs]: Can we do something like this here for some additional safety before casting to long:
```csharp
var rawTicks = _settings.RetryInterval.Ticks * backoffFactor * (1.0 - jitter);
return TimeSpan.FromTicks(
        (long)Math.Min(rawTicks, _settings.MaxRetryInterval.Ticks)
);
```

**Review by gregyjames — APPROVED** [review]
LGTM! Sorry for the delay in approving and all the requested changes. This release will be a huge milestone for the project, very much appreciate all the hard work on this.

File: src/LeaderElection/LeaderElectionBase.cs (hunk 15 above)
Comment (mrfootoyou) [reply src/LeaderElection/LeaderElectionBase.cs]: 👍 Done.
